src/mpl2/src/hier_rtlmp.cpp

```diff
@@ -914,142 +912,119 @@ void HierRTLMP::setTightPackingTilings(Cluster* macro_array)
   macro_array->setMacroTilings(tight_packing_tilings);
 }
 
-void HierRTLMP::setIOClustersBlockages()
+void HierRTLMP::setPinAccessBlockages()
 {
   if (!tree_->maps.bterm_to_inst.empty()) {
     return;
   }
 
-  IOSpans io_spans = computeIOSpans();
-  const float depth = computeIOBlockagesDepth(io_spans);
+  std::vector<Cluster*> io_clusters = getIOClusters();
+  const Rect die = dbuToMicrons(block_->getDieArea());
 
-  const Rect root(tree_->root->getX(),
-                  tree_->root->getY(),
-                  tree_->root->getX() + tree_->root->getWidth(),
-                  tree_->root->getY() + tree_->root->getHeight());
+  const float depth = computePinAccessBlockagesDepth(io_clusters, die);
+
+  for (Cluster* io_cluster : io_clusters) {
+    Boundary constraint_boundary = io_cluster->getConstraintBoundary();
+    createPinAccessBlockage(constraint_boundary, depth, die);
+  }
 
-  // Note that the range can be larger than the respective core dimension.
-  // As SA only sees what is inside its current outline, this is not a problem.
-  if (io_spans[L].second > io_spans[L].first) {
-    const Rect left_io_blockage(root.xMin(),
-                                io_spans[L].first,
-                                root.xMin() + depth,
-                                io_spans[L].second);
+  if (boundary_to_io_blockage_.empty()) {
+    // If there are no constraints at all, give freedom to SA so it
+    // doesn't have to deal with pin access blockages in all boundaries.
+    // This will help SA not relying on extreme utilizations to
+    // converge for designs such as sky130hd/uW.
+    if (tree_->blocked_boundaries.empty()) {
+      return;
+    }
+
+    // There are only -exclude constraints, so we create pin access
+    // blockages based on the boundaries that are not blocked.
+    if (tree_->blocked_boundaries.find(L) == tree_->blocked_boundaries.end()) {
+      createPinAccessBlockage(L, depth, die);
+    }
+
+    if (tree_->blocked_boundaries.find(R) == tree_->blocked_boundaries.end()) {
+      createPinAccessBlockage(R, depth, die);
+    }
 
+    if (tree_->blocked_boundaries.find(B) == tree_->blocked_boundaries.end()) {
+      createPinAccessBlockage(B, depth, die);
+    }
+
+    if (tree_->blocked_boundaries.find(T) == tree_->blocked_boundaries.end()) {
+      createPinAccessBlockage(T, depth, die);
+    }
+  }
+}
+
+void HierRTLMP::createPinAccessBlockage(Boundary constraint_boundary,
+                                        const float depth,
+                                        const Rect& die)
+{
+  if (constraint_boundary == L) {
+    const Rect left_io_blockage(
```

Comment:
Similar to another comment I made, in the future we should work on enhancing this process to consider the actual constraint region. Blocking the whole edge might be to extreme for cases where the constraint region is small.

Reply:
I''ll open an issue for that.

Reply:
Opened #5816